soaktest/soaktest.go

```diff
@@ -0,0 +1,258 @@
+		}
+
+		rttValue := broker.(map[string]interface{})["rtt"]
+		avg, ok := rttValue.(map[string]interface{})["avg"].(float64)
+		if !ok {
+			ErrorLogger.Printf("Failed to convert avg to float64 %v\n", ok)
+			return ok
+		}
+
+		p99, ok = rttValue.(map[string]interface{})["p99"].(float64)
+		if !ok {
+			ErrorLogger.Printf("Failed to convert p99 to float64 %v\n", ok)
+			return ok
+		}
+
+		nid, ok := broker.(map[string]interface{})["nodeid"].(float64)
+		if !ok {
+			ErrorLogger.Printf("Failed to convert nodeid to float64 %v\n", ok)
+			return ok
+		}
+		nodeid = int(nid)
+
+		tags := []string{fmt.Sprintf("broker:%d type: %s topic: %s parts: %s",
+			nodeid, raw["type"], topic, parts)}
+
+		DatadogGauge(BROKER_RTT_P99, p99/1000000.0, tags)
+		DatadogGauge(BROKER_RTT_AVG, avg/1000000.0, tags)
+	}
+	return true
+}
+
+// HandleStatsEvent converts Stats events to hashmap
+func HandleStatsEvent(e *kafka.Stats, topic string) error {
+	var raw map[string]interface{}
+	err := json.Unmarshal([]byte(e.String()), &raw)
+	if err != nil {
+		ErrorLogger.Printf("Json unmarshall error: %s\n", err)
+		return err
+	}
+
+	ok := rttStats(raw, topic)
+	if !ok {
+		return fmt.Errorf("Failed to calculate broker rtt statistics")
+	}
+	return nil
+}
+
+// ProducerDeliveryCheck handles delivery report for producer
+func ProducerDeliveryCheck(e *kafka.Message) {
+	tags := []string{fmt.Sprintf("topic: %s", *e.TopicPartition.Topic)}
+	if e.TopicPartition.Error != nil {
+		DrErrCnt += 1
+		DatadogIncrement(PRODUCER_DR_ERR, 1, tags)
+		ErrorLogger.Printf("Delivery failed: %v\n", e.TopicPartition)
+	} else {
+		DrCnt += 1
+		DatadogIncrement(PRODUCER_DR, 1, tags)
+		if DrCnt%1000 == 0 {
+			InfoLogger.Printf("Delivered message to topic at events channel "+
+				"%s [%d] at offset %d\n", *e.TopicPartition.Topic,
+				e.TopicPartition.Partition, e.TopicPartition.Offset)
+		}
+	}
+}
+
+func PrintConsumerStatus() {
+	InfoLogger.Printf("Consumer: %d messages consumed, %d duplicates, "+
+		"%d missed, %d message errors, %d consumer client-level errors\n",
+		ConsumerMsgCnt, MsgDupCntForOutput, MsgMissCntForOutput,
+		ConsumerMsgErrs, ConsumerErr)
+}
+
+func PrintProducerStatus() {
+	InfoLogger.Printf("Producer: %d messages produced, %d delivered, %d "+
+		"failed to deliver, %d producer client-level errors\n",
+		ProducerMsgId, DrCnt, DrErrCnt, ProducerErr)
+}
+
+// HandleMessage handles received messages, monitors latency and
+// verifies messages
+func HandleMessage(e *kafka.Message,
+	hwmarks map[string]uint64) bool {
+	tags := []string{fmt.Sprintf("topic: %s", *e.TopicPartition.Topic)}
+	if e.TopicPartition.Error != nil {
+		ConsumerMsgErrs += 1
+		DatadogIncrement(CONSUMER_RECEIVE_ERROR, 1, tags)
+		ErrorLogger.Printf("Message receive failed from: %s\n",
+			e.TopicPartition.Error)
+		return false
+	}
+	ConsumerMsgCnt += 1
+	DatadogIncrement(CONSUMER_CONSUME_MSG, 1, tags)
+	if ConsumerMsgCnt%1000 == 0 {
+		InfoLogger.Printf("Consumer out Message on %s:\n%s\n%s\n",
```

Comment:
Multiline logs should be avoided if possible since they make grepping and sorting cumbersome.